### DISCUSSION TITLE
Use with numpy>=2.0 fails to cast np.float64() into SQL correctly?

### DISCUSSION THREAD

**Discussion by falloutphil** [sqlalchemy/sqlalchemy#11712] [Usage Questions] [answered]
Upgrading to numpy 2.0, I see errors when trying to use Postgres with SQLAlchemy - there are no issue with Sqlite after upgrade.

The problem seems to be numpy floats are being written into the SQL statement - one example of this is before but this happens in many places in the codebase I work on (codebase is proprietry so can't give details here but will try to produce toy example).

I suspect that values returned from np.sum or similar were previously cast automatically to a type that inserted into the SQL, but now that's not working.

Anyone seen this - is the workaround to cast everything to python int?  Any ideas why previously this worked implicitly?

```SQL
 (psycopg2.errors.InvalidSchemaName) schema "np" does not exist
LINE 1: ...eak_tenor_id, gamma, vega, rega, sega) VALUES (1, np.float64...
                                                             ^
[SQL: INSERT INTO std_superior_risk_metric (id, superior_risk_value, superior_risk_id, peak_tenor_id, gamma, vega, rega, sega) VALUES (%(id)s, %(superior_risk_value)s, %(superior_risk_id)s, %(peak_tenor_id)s, %(gamma)s, %(vega)s, %(rega)s, %(sega)s)]
[parameters: {'id': 1, 'superior_risk_value': np.float64(143.06439161545444), 'superior_risk_id': 1, 'peak_tenor_id': 1, 'gamma': np.float64(143.06439161545444), 'vega': np.float64(0.0444985191510483), 'rega': np.float64(0.0), 'sega': np.float64(0.0)}]
(Background on this error at: https://sqlalche.me/e/20/f405)
``` 

Downgrading back to final v1 numpy release from Feb 2024 (1.26.4) and the problem goes away.
 
Thanks!

**Comment by CaselIT** [comment]
Hi,

This is likely a question for psycopg2.
You probably need to tell it to how to convert numpy floats using https://www.psycopg.org/docs/advanced.html#adapting-new-python-types-to-sql-syntax

**Reply by kjyv** [reply]
Could this also be fixed possibly by using psycopg3?

**Reply by CaselIT** [reply]
there are similar features in psycopg3.

3.2 should support it out of the box https://www.psycopg.org/psycopg3/docs/news.html#psycopg-3-2

**Comment by falloutphil** [comment, marked as answer]
The move and use of 3.2 as suggested by @CaselIT is a good one - and this certainly handles any np.float64s passed directly into SQLAlchemy statements.  My problem seems to have been the use of `repr()` prior to that in the execution stack, meaning the damage was done casting the float64 to a string before psycopg had a chance to handle it properly.

The good news is there is a quick workaround which allows you to use numpy v2 without having to fix all the issues up front - just make sure you import numpy and set this as early as possiblty in your application:

```python
import numpy as np
np.set_printoptions(legacy="1.25")
```

See here for more details: https://numpy.org/doc/stable/reference/generated/numpy.set_printoptions.html
